### ISSUE TITLE
FYI: how we handle "releases"

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#54]
EDITED: 15-03-2017

"releases" are end of stable-life-cycle for archival/legacy purposes (with all the benefits of being fully up-to-date with any changes including vetting/testing/feedback over the life-cycle, as well as typos, extra items, revamps, whatever).

"pre-releases" - An important milestone for end users is knowing when the master becomes "version compliant" - and that point is when we have finished all the diffs generated by earthlng's diff script. The closer we are to when that version lands from Mozilla, the better. To let users know (because a closed "ToDo: Diff" issue doesn't necessarily cut it), we have the changelog (a more user friendly version of what's happening as compared to an atomic level commit history). This is also the basis for ghacks doing a 6 weekly article as a follow up to his 6 weekly "What's new in FFxx". This is good for feedback, knowledge for end-users, and also drawing contributors and users here. In order for earthlng to generate the changelog, he needs two points. And for all the reasons you would expect, having those snapshot points stored here makes sense. But we will use a "pre-release" tag.

Note, we also have a `changelog` label, so people can use `+label%3Achangelog` to pull up a running history (although we will have only one open, and a backlog of closed ones).

--typical timeline (example 53->54)--

*  -x days
   * we have 53->54 diffs issue in advance, we know a lot of what will be coming, and can prepare etc
*  -4 or -2 or -1 days:
   * create a 53 release for archive/legacy
   * change master to version 54-beta, set new date, set new code name & lyric
   * start commits for 54
* D-Day
   * 54 stable lands from Mozilla
* +4 days or +10 days (i.e *when* we have finished the 53->54 diffs)
   * close `diffs 53->54` issue
   * update `date` in master and change to version 54
   * create a `54-alpha` release
   * generate a diff (`53-alpha`->`54-alpha`)
   * from the above diff, create a `changelog: 54-alpha` as an open issue
   * close `changelog: 53-alpha` issue
   * ghacks does an article a day or two later
* repeat cycle

Throughout the entire life-cycle, we will make changes as required. Typos, new reference sources, revamps, changing items active status, and so on.

Note: The "release" and "alpha release" descriptions make it clear what they are. If end users choose to download the "alpha release" zip, even weeks after it has been created, then that's their problem. The master is always current stable, and always being improved. It will however, become a bit more of a moot point as this js reaches finalization and there's nothing more to do except version changes - .ie, nothing will change for weeks and weeks on end.

**Comment by Just-me-ghacks** [comment]
I don't think we have the manpower for "a release to coincide with Firefox stable landing". For the latest version I would always go for the "current up-to-date live js" anyway. My vote goes for a high quality legacy archived version. A changelog (as [the one suggested by earthlng](https://github.com/ghacksuserjs/ghacks-user.js/issues/53)) would be handy too. I love the format and the style.

**Comment by earthlng** [comment]
Man, Pants, if there have ever been two people that probably shouldn't work together - it's the two of us.
We argue about everything, and even about the dumbest things - for hours :)

Look, I don't think my idea has any problems. We can just make it abundantly clear that "legacy users" should use a "legacy version" - there, problem solved.
And the other point - that we maybe won't always have the time or willingness to have a new release ready when a new stable FF is released or soon thereafter - is also no problem in my opinion. So what, it may take a while until we've looked at the diffs and decided on what to do with the new/changed prefs, but as soon as we've done it, people will see the new release and know that what we think should be included for FF52 is now there and ready to use. If they don't want to wait they can look at our diffs and decide and include/remove prefs themselves in their local copy.
In the changelog I tried to point out all the prefs that could cause any problems for end-users.
This time many changes only enforce the default value now anyway, so they are safe to use no matter what. And if someone wants to be on the safe side, he/she can just ignore the potentially problematic ones.

As soon as we have re-vamped the remaining sections, done all the re-ordering and re-numbering, we won't even necessarily have to do a changelog anymore. The commits can then be used as the changelog instead. I mean we have 400+ prefs - at a certain point there's not much more that we need to change.

I think it's also safe to say that there is probably not a single person who uses this AS IS, anyway.
One guy on reddit even said "You're not **supposed** to use it as is" (admittedly a bit harsh considering what we're trying to achieve here)
So who exactly are we trying to please here?
Ok, I've said my peace, I don't mind if we do it your way. Let's see what the other guys think and then it's up to you Pants to decide what we'll do.
I have no problem with rolling over + being the bottom-bitch. I'll do that more often from now on, so we won't waste as much valuable time anymore with arguing over stupid things.

**Comment by atomGit** [comment]
i just read the 'todo: release v52' post

i think i can sorta understand the conflict regarding when to release what - i sure don't wish to add fuel to the fire, but my personal idea would be to release a version ASAP that coincides with the latest FF release - so moz releases v53 and you release v53 shortly thereafter and it becomes a work-in-progress

when moz blows out v54, then your v53 gets dumped into a v53 folder and basically abandoned unless someone contributes (finds a bug, etc.) in which case you release a new build if the bugh or whatever is important enough - this is why i'm releasing mine as v52r1, r2, etc., so the major version is tied to the major version of FF and the users can differentiate between revisions for a particular version and thus can always be running my latest revision

the problem (if you want to call it that) that i see with the current method is that those running the latest _release_ of your user.js will forever be behind the 8-ball because they won't be leveraging the latest and greatest changes since there won't be a new release until the next major version of the browser which isn't optimal IMO from a security/privacy standpoint

please don't let this start any arguments - this is only my personal take and this project is too important to get shit-canned - i understand the passion though

**Comment by Atavic** [comment]
### As an archive legacy zip :+1: 
IMHO

@Roman-Nopantski wisely said
> The LIVE user.js is always in a state of flux.

Latest release should simply be the same here and on Ghacks articles.
There's so little people for a bleeding-edge approach here.

**Comment by Roman-Nopantski** [comment]
Butting heads is good. No one is right all the time. Or there could be two+ ways to achieve the same result. I think we are all smart enough to actually "listen" to the other person and make rational decisions. I would rather be questioned than surrounded by yes men. As I've always said:

"Walk a mile in the other person's shoes. *Then* criticize them. One, you'll be a mile away, and two, you'll have their shoes!"

--
Maybe you're not parsing what I said, or I am cr*p at explaining.

> I don't think my idea has any problems. We can just make it abundantly clear that "legacy users" should use a "legacy version" - there, problem solved.

But the zipped up versions sitting in the "releases" section are the legacy versions, and they would be created at the start of their life, not the end.

> And the other point - that we maybe won't always have the time or willingness to have a new release ready when a new stable FF is released or soon thereafter

I never said that. We archive off the previous version (say 3 or 4 days before end of life), and immediately change the master js numbering by one, new code name and date. NOW we have a live new 6 week cycle of the new stable. In our OWN time, we do the diffs, and when WE are ready, Martin does an article. Simple huh.

> As soon as we have re-vamped the remaining sections, done all..

Indeed. I want most stuff done and dusted before Martin's launch article. So in future, users and forks have "clean" relevant minor commits. This is also another reason to keep piling on the changes between 51 "release" and 52 "release" .. so that 52 is as close as possible to all cleaned up.

**Comment by Roman-Nopantski** [comment]
@all - I am pretty muck AFK (from here) for the next 48 hours. Gotta pay the bills.

**Comment by publicarray** [comment]
I also agree master should target the latest version of FF and may even be unstable. I just wanted to add that with git and GitHub you can Tag a releases. So you could create a tag just before a new version of FF is targeted (i.e. before new preferences are added for the next version of FF). In the tag it should be clear what the version of FF it was developed for. It would make it easy to search for a tag, version or release that works for any given FF version from now onwards.
The only disadvantage I can think of is that older versions are no longer maintained (but would that be problem?)

**Comment by earthlng** [comment]
hi, thanks for your input. the master will always target the latest version of FF anyway, that's out of question.
Pants and myself didn't agree on what our "release" should be. But I think we now found an acceptable solution for both us. That is, we try to be ready with all the pref-changes for the next FF release on the day (or soon after) that the new FF stable is released and we'll then create a changelog and release our "earthlng user.js v53" (for example). Although Pants would prefer that people don't use that version, everyone is free to do as they wish.

> Branches can also help in segregating different targets.

Yes, I also considered using branches for different versions. It would allow us to "back-port" changes if necessary, but IMO it would be a PITA and most likely quickly become unmanageable, confusing and turn into a mess.
We'll see how it goes with our current approach and maybe adjust when necessary.

**Comment by Thorin-Oakenpants** [comment]
This is all sorted out.

"releases" are end of stable-life-cycle for archival/legacy purposes (with all the benefits of being fully up-to-date with any changes including vetting/testing/feedback over the life-cycle, as well as typos, extra items, revamps, whatever).

"pre-releases" - An important milestone for end users is knowing when the master becomes "version compliant" - and that point is when we have finished all the diffs generated by earthlng's diff script. The closer we are to when that version lands from Mozilla, the better. To let users know (because a closed "ToDo: Diff" issue doesn't necessarily cut it), we have the changelog (a more user friendly version of what's happening as compared to an atomic level commit history). This is also the basis for ghacks doing a 6 weekly article as a follow up to his 6 weekly "What's new in FFxx". This is good for feedback, knowledge for end-users, and also drawing contributors and users here. In order for earthlng to generate the changelog, he needs two points. And for all the reasons you would expect, having it those snapshot points stored here makes sense. But we will use a "pre-release" tag.

Note, we also have a `changelog` label, so people can use `+label%3Achangelog` to pull up a running history (although we will have only one open, and a backlog of closed ones).

--typical timeline (example 53->54)--

*  -x days
   * we have 53->54 diffs issue in advance, we know a lot of what will be coming, and can prepare etc
*  -4 or -2 or -1 days:
   * create a 53 release for archive/legacy
   * change master to version 54, set new date, set new code name & lyric
   * start commits for 54
* D-Day
   * 54 stable lands from Mozilla
* +4 days or +10 days (i.e *when* we have finished the 53->54 diffs)
   * close `diffs 53->54` issue
   * update `date` in master
   * create a `pre-release 54`
   * generate a diff (`pre-release53`->`pre-release54`)
   * `pre-release53` is now no longer needed, and can be deleted
   * from the above diff, create a `changelog: 54` as an open issue
   * close `changelog: 53` issue
* repeat cycle

Throughout the entire life-cycle, we will make changes as required. Typos, new reference sources, revamps, changing items active status, and so on.

Note: The "release" and "pre-release" descriptions make it clear what they are. If end users choose to download the "pre-release" zip, even weeks after it has been created, then that's their problem. The master is always current stable, and always being improved. It will however, become a bit more of a moot point as this js reaches finalization and there's nothing more to do except version changes - .ie, nothing will change for weeks and weeks on end.

**Comment by earthlng** [comment]
We could really use some more help here. We have the changelog now, so can't you ask Martin to do an article already? I mean you wrote

> This is also the basis for ghacks doing a 6 weekly article as a follow up to his 6 weekly "What's new in FFxx". This is good for feedback, knowledge for end-users, and also drawing contributors and users here.

I very much doubt that we will get any more contributors, but as they say "hope dies last", so we should at least try. Otherwise the current changelog is basically useless, right?

**Comment by Thorin-Oakenpants** [comment]
The current changelog is not useless. And neither will future changelogs as they become old. They are historical.

I have a timeline and plan. I already have some drafts here for the three wikis pages i need, and the readme will be easy (and short as f**k). *When* I am ready, I will give Martin an article - It will be timed nicely so its spaced well before before his next "Whats in FFxx". This article to let everyone know to come here, will include the major changes we have done so far. It will mention the changelogs and how we aim to provide an alpha as fast as possible. Rome wasn't built in a day :) Its not like I've been sitting on my ass :)

I am semi-assuming Martin will do a 6 weekly article. I am 90% sure he will, as when I approached him about going github he said he could let people know as I did releases. That we will do 6 weekly ones probably didn't cross his mind. But that said, all he has to do is output your changelog (and you named as author - yay, you're famous!), so I see no issues. What we gain from it, who knows. Initially lots, later on very little. At least it might spark some debate from users not willing to login to github. And attracting people like "Parker Lewis" on board here would awesome. I also see growth after 57, when WE can't change prefs that are not their own. "The rise of the user.js: judgment day"

Anyway, the first changelog is done. Its relevant, and will still be front and center until the next one (5 more weeks?). I hope to have an article in about 10 days.
